sph/include/sph/positions.hpp

```diff
@@ -42,21 +42,90 @@
 
 namespace sph
 {
+//! @brief checks whether a particle is close to a fixed boundary and reduces a scalar if so
+template<class Tc, class Th, class Ts>
+HOST_DEVICE_FUN void fbcAdjustScalar(const cstone::Vec3<Tc> X, Ts& scalar, const cstone::Box<Tc>& box, const Th hi)
+{
+    // half the width of the sigmoid, in smoothing lengths
+    constexpr Th width    = 2.;
+    constexpr Th invWidth = 1 / width;
+    // width of particles that are held at zero, in smoothing lengths
+    constexpr Th       offset          = 2.;
+    cstone::Vec3<bool> isBoundaryFixed = {
+        box.boundaryX() == cstone::BoundaryType::fixed,
+        box.boundaryY() == cstone::BoundaryType::fixed,
+        box.boundaryZ() == cstone::BoundaryType::fixed,
+    };
+    cstone::Vec3<Tc> boxMax = {box.xmax(), box.ymax(), box.zmax()};
+    cstone::Vec3<Tc> boxMin = {box.xmin(), box.ymin(), box.zmin()};
+
+    for (int j = 0; j < 3; ++j)
+    {
+        if (isBoundaryFixed[j])
+        {
+
+            Th relDistanceMax = std::abs(boxMax[j] - X[j]) / hi;
+            Th relDistanceMin = std::abs(boxMin[j] - X[j]) / hi;
+            Th minDistance    = relDistanceMin < relDistanceMax ? relDistanceMin : relDistanceMax;
+
+            if (minDistance < 2 * width + offset)
+            {
+                Tc correction = 0.5 * (std::tanh(4 * (minDistance - offset) * invWidth - 4) + 1);
+                scalar *= correction;
+            }
+        }
+    }
+}
 
-//! @brief checks whether a particle is in the fixed boundary region in one dimension
+//! @brief checks whether a particle is close to a fixed boundary and reduces the acceleration if so
 template<class Tc, class Th>
-HOST_DEVICE_FUN bool fbcCheck(Tc coord, Th h, Tc top, Tc bottom, bool fbc)
+HOST_DEVICE_FUN void fbcAdjustVector(const cstone::Vec3<Tc> X, cstone::Vec3<Tc>& V_nm, cstone::Vec3<Tc>& A,
+                                     const cstone::Box<Tc>& box, const Th hi)
```

Comment:
This is a a) duplicate of the scalar version, which could be re-used in here by calling component by component and b) won't be needed if you compute the correction factor once for the particle and apply the factor where needed.